### ISSUE TITLE
enhancement: show the back trace in case of a thrown exception in debug mode

### ISSUE THREAD

**Issue by jakobkroeker** [Macaulay2/M2#585]
Is it currently possible to geto rshow the back trace in case of a thrown exception ?
(`throw "exception"`)
If not, I find that could be useful for debugging.

**Comment by DanGrayson** [comment]
Throwing something is not an error, so there would be no occasion to show a backtrace.  The whole point is to return to a higher level, where the "catch" was, without annoying the user.

**Comment by jakobkroeker** [comment]
throwing something is not an error but a user might have been used a method incorrectly and he did not expect the exception and might want to figure out quickly what is going on.
Other languages: in C# the stack trace is provided by default in case of an exception.

So I'm not yet convinced.

**Comment by DanGrayson** [comment]
Please say what you mean by "exception" in the context of Macaulay2.  Do you mean "error"?

**Comment by jakobkroeker** [comment]
No, I do not mean "error", I mean "throw ..."

**Comment by DanGrayson** [comment]
Do you mean you would like to be able to set a global flag so that all subsequent uses of "throw" would result in backtraces of the stack frames discarded?  That seems too special, since the use of "throw" should not be consider an exception -- it's just a normal way to transfer control, similar to "break" or "return" or "continue".

**Comment by jakobkroeker** [comment]
It seems we have different ideas for "throw" concept in M2
My view/wish of 'exception' is an catcheable error distinguishable by error type and distinguishable from regular return value type. So an Error would be a type, lets say derived from HashTable where one of the entries is the back trace (enabled or disabled by default)

**Comment by DanGrayson** [comment]
Yes, it would be very nice if various types of such "exceptions" could be distinguished from each other, so that different uses of "throw" could be caught by different uses of "catch".  But exceptions are not like errors, and need not be brought to the attention of the user, by backtrace or otherwise.  If you have an error situation, don't use "throw".

**Comment by jakobkroeker** [comment]
> But exceptions are not like errors, and need not be brought to the attention of the user, by backtrace or 
otherwise. 


I guess I was not precise enought. *sigh*
Are we talking about a caught or uncaught exceptions?
I find that the user or developer might be interested in the back trace if an exception was _not catched_
It might also be that we have a different opinion on this. 

Therefore I predict that after M2 package developers will have been used  `throw..catch` for a while the M2 language will converge to provide back trace at least for uncaught exceptions.

**Comment by DanGrayson** [comment]
I thought you were interested in any use of "throw".

Yes, certainly it would be an error to use "throw" without also using the corresponding "catch", but it would
slow down Macaulay2 to save backtrace information for every "throw" just in case it needs to be printed out if a "throw" is not found by unwinding the stack all the way.  But you make a good point -- if the code containing "throw" is also getting used correctly by code containing "catch", then it can be hard to locate the code not containing "throw" that calls it incorrectly, unless a backtrace is available.

I don't have a good recommendation.

**Comment by jakobkroeker** [comment]
> I don't have a good recommendation.

my first thought:
what about opt-in collecting backtrace? (disabled by default)

**Comment by mahrud** [comment]
@jakobkroeker in case you're still interested in this, `stacktrace()` is now an experimental function of the interpreter that returns a C++ stacktrace. If you compile with `-DPROFILING=1` then you get a stacktrace of the top level (not a great design, I know, I should have started with setting some design parameters).

**Comment by d-torrance** [comment]
> **Written by Claude** (Claude Opus 5, via Claude Code), not by @d-torrance, whose account posted it -- please weigh it accordingly.

Closing as not planned, from a sweep of the open issues — not because the request was
answered, but because nine years on the situation it protects against has barely arisen.

### The scenario needs `throw` to be in use, and it is not

The case @DanGrayson eventually granted here is a real one:

> if the code containing "throw" is also getting used correctly by code containing "catch",
> then it can be hard to locate the code not containing "throw" that calls it incorrectly,
> unless a backtrace is available.

That requires packages to use `throw`/`catch` as control flow. Across the whole package
tree today, essentially one does — `DirectSummands.m2`, at four call sites:

```m2
if isSurjective c then throw (c, rightInverse c)              -- :538
if isIsomorphism(c * b) then throw (c, b)                     -- :548, :555, :560
```

Every other match is prose: comments like "otherwise, throw error", documentation saying a
method "will throw an error", and `HardDegreeLimit => {}  -- throw out information in
degrees above this one`. Core uses `throw` once.

So the cost @DanGrayson objected to — *"it would slow down Macaulay2 to save backtrace
information for every throw just in case"* — is constant and real, and the benefit is very
nearly hypothetical.

### On `stacktrace`

@mahrud's 2020 note is adjacent to the request rather than an answer to it. `stacktrace`
takes an integer and has two modes, neither of which is an automatic backtrace at an
uncaught throw:

- `stacktrace 0` prints a **C++** trace via Boost::stacktrace — `evaluate_evalraw`,
  `readeval3`, engine frames with `.dd` line numbers. Useful for debugging the interpreter,
  not for finding the M2 code that called something wrongly.
- `stacktrace n` for `n > 0` returns a sequence of positions from `M2_stack`, the profiler's
  top-level call log, and
  [`d/profiler.dd:26`](https://github.com/Macaulay2/M2/blob/development/M2/Macaulay2/d/profiler.dd#L26)
  still carries `-- TODO: limit this to Code, packages, stc for traceDepth = 1, 2, 3, etc.`,
  so the depth argument is not really implemented.

Both are things you call deliberately, at a point you have already identified — which is the
opposite of the situation this issue is about.

### If it is ever wanted

@jakobkroeker's last suggestion here was never answered and is still the obvious shape:
opt-in collection, disabled by default. That answers the performance objection directly,
since nothing is saved unless somebody asks for it. It is recorded here so that whoever
needs it does not have to rediscover it — and if `throw`/`catch` ever becomes a common
idiom in packages, this should be reopened rather than refiled.

Closing on lack of demand, not on the merits. Thanks for the careful argument at the time,
@jakobkroeker — the use case you identified was granted, it simply has not materialised.

<sub>Triaged in a sweep of the open issues; source of truth: [`bug-triage/issues.tsv`](https://github.com/d-torrance/M2/blob/bug-triage/bug-triage/issues.tsv).</sub>
